Sources/StreamChat/Controllers/DatabaseObserver/StateLayerListDatabaseObserver.swift

```diff
@@ -0,0 +1,111 @@
+//
+// Copyright © 2024 Stream.io Inc. All rights reserved.
+//
+
+import CoreData
+import Foundation
+
+/// A CoreDate store observer which immediately reports changes as soon as the store has been changed.
+///
+/// List changes are reported using item ids to reduce the burden of converting DTOs to models.
+@available(iOS 13.0, *)
+final class StateLayerListDatabaseObserver<Item, ItemID, DTO: NSManagedObject> {
```

Comment:
While writing tests I encountered an issue where this did not work:
```swift
try await channelList.loadChannels(with: pagination)
let channels = channelList.state.channels // <-- new channels were not returned
```

The core issue was that if I change the local store using `writableContext` but I observe the store using the `backgroundContext`, then there is a slight delay before `backgroundContext` is notified about the store change. Moreover, `BackgroundListDatabaseObserver` what I was using before has internal operation queue which delays the change even more. I also looked into using `ListDatabaseObserver` but it does some unnecessary model conversions the state layer does not need (see ChangeAggregator which converts DTOs to models). Also there was an opportunity to skip `performAndWait` and `@Atomic` locking when FRC reported changes. Thirdly, I was able to hide bring in async callback to this level and do the StreamCollection conversion here instead.
The cost is that we have another database observer.

Reply:
ok, let's merge it in this state, but add it to the list. I think we should try a separate context.

Reply:
I'll add it as a separate TODO and will try it in a separate PR. If it does not work, tests will fail immediately.